data_steward/tools/snapshot_by_query.py

```diff
@@ -37,33 +37,64 @@ def create_empty_cdm_tables(snapshot_dataset_id):
     cdm.create_vocabulary_tables(snapshot_dataset_id)
 
 
-def get_field_cast_expr(field, data_type):
+def get_field_cast_expr(field, data_type, required_mode, existing_fields):
     """
     generates cast expression based on data_type for the field
 
     :param field: field name
     :param data_type: data type of the field
+    :param required_mode: if a field is required or nullable
+    :param existing_fields: list of field names in a table
     :return:  col string
     """
-    bigquery_int_float = {'integer': 'INT64', 'float': 'FLOAT64'}
+    bigquery_data_types = {
+        'integer': 'INT64',
+        'float': 'FLOAT64',
+        'string': 'STRING',
+        'date': 'DATE',
+        'timestamp': 'TIMESTAMP'
+    }
+    replacement_values = {
+        'integer': 0,
+        'float': 0.0,
+        'string': '\'\'',
+        'date': 'DATE(\'1970-01-01\')',
+        'timestamp': 'TIMESTAMP(\'1970-01-01\')'
+    }
+    if field not in existing_fields and required_mode == 'required':
+        col = f'CAST({replacement_values[data_type.lower()]} AS {bigquery_data_types[data_type.lower()]}) AS {field}'
+    elif field not in existing_fields and required_mode == 'nullable':
+        col = f'CAST(NULL as {bigquery_data_types[data_type.lower()]}) AS {field}'
+    elif required_mode == 'required':
+        col = f'CAST(COALESCE(CAST({field} AS {bigquery_data_types[data_type.lower()]}), {replacement_values[data_type.lower()]}) AS {bigquery_data_types[data_type.lower()]}) AS {field}'
+    elif required_mode == 'nullable':
+        col = f'CAST(COALESCE(CAST({field} AS {bigquery_data_types[data_type.lower()]}), NULL) AS {bigquery_data_types[data_type.lower()]}) AS {field}'
+    return col
 
-    if data_type not in ['integer', 'float']:
-        col = f'CAST({field} AS {data_type.upper()}) AS {field}'
-    else:
-        col = f'CAST({field} AS {bigquery_int_float[data_type]}) AS {field}'
 
-    return col
+def get_existing_field_names(table_id, dataset_id):
+    """
+    gets fields present in table_id in the input dataset
+    :param table_id: name of the table to gets field names
+    :param dataset_id: name of the input dataset
+    returns a list of field names
+    """
+    return [
+        field['name']
+        for field in get_table_info(table_id, dataset_id)['schema']['fields']
+    ]
 
 
 def get_copy_table_query(project_id, dataset_id, table_id):
     try:
         table_name = table_id
         fields = resources.fields_for(table_name)
-        fields_with_datatypes = [
-            get_field_cast_expr(field['name'], field['type'])
-            for field in fields
+        existing_field_names = get_existing_field_names(table_id, dataset_id)
+        col_cast_exprs = [
+            get_field_cast_expr(field['name'], field['type'], field['mode'],
+                                existing_field_names) for field in fields
```

Comment:
As discussed lets use the standard bq client rather than the `bq_utils`. I think it'd be a bit easier to follow too: 

```suggestion
        client = bq.get_client(project_id)
        table = client.get_table(f'{dataset_id}.{project_id}')
        dst_fields = resources.fields_for(table_id)
        col_cast_exprs = (get_field_cast_expr(field, table.schema) for field in dst_fields)
```